### PR TITLE
K8S version upgrade v0.29.x and related changes

### PR DESCRIPTION
**Pull request by mabhi**
## Change Overview

The following PR is the revert of previous reverted PR: #2676 .
This PR includes :
- Restore changes on k8s upgrade version v0.29.x
- A change wherein validation webhook being created is added explicitly to the controller manager. 
Without the second change there was always a chance that the mutating/validating webhook request never reached the controller
- New validation methods added to service validation webhook requests.
 
## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [x] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues 

- fixes #issue-number

## Test Plan

- [x] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

Install kanister controller using private image in repo with local changes (TAG=v0.0.6)
`helm upgrade --install kanister ./helm/kanister-operator --create-namespace -n kanister --set image.repository=mabhi2ic/controller --set image.tag=${TAG}`

Check the correct version of the controller image with the changes
```
(⎈|kind-local:default)➜  kanister git:(revert-k8s-upgrade-v0-29-x) ✗ k describe -n kanister po kanister-kanister-operator-7c8bf8766-kd8lp

Name:             kanister-kanister-operator-7c8bf8766-kd8lp
Namespace:        kanister
....
Status:           Running
IP:               10.244.0.16
......
Containers:
  kanister-kanister-operator:
    Container ID:   containerd://f21dab87ebdd6515c5b2261e5db81aaf533fbccb5312ea4e58d05ae68debfacb
    Image:          mabhi2ic/controller:v0.0.6
     .....
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-xlc5l (ro)
      /var/run/webhook/serving-cert from webhook-certs (rw)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  webhook-certs:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  kanister-webhook-certs
    Optional:    false
```
Testing by issuing a valid blueprint
```
(⎈|kind-local:default)➜  kanister git:(revert-k8s-upgrade-v0-29-x) k get po -n kanister   
NAME                                          READY   STATUS    RESTARTS   AGE
kanister-kanister-operator-7df894bb95-gmlmr   1/1     Running   0          29s
(⎈|kind-local:default)➜  kanister git:(revert-k8s-upgrade-v0-29-x) k create -f ./examples/mysql/mysql-blueprint.yaml -n kanister
blueprint.cr.kanister.io/mysql-blueprint created
```
Container logs for the above -
```
-------------------------------------------------------------------------
Passed the 'validation of phase restoreFromBlobStore in action restore' check.. ✅
Passed the 'validation of phase dumpToObjectStore in action backup' check.. ✅
Passed the 'validation of phase deleteFromBlobStore in action delete' check.. ✅
{"level":"info","ts":"2024-02-20T18:40:10Z","logger":"admission","msg":"Validating blueprint","object":{"name":"mysql-blueprint","namespace":"kanister"},"namespace":"kanister","name":"mysql-blueprint","resource":{"group":"cr.kanister.io","version":"v1alpha1","resource":"blueprints"},"user":"kubernetes-admin","requestID":"5d22fff4-4ec5-4719-b24e-b1212935f35b"}
{"File":"pkg/controller/controller.go","Function":"github.com/kanisterio/kanister/pkg/controller.(*Controller).onAddBlueprint","Line":231,"cluster_name":"94c9c8a4-f51d-46eb-a15c-019e6d6beabc","hostname":"kanister-kanister-operator-7df894bb95-gmlmr","level":"info","msg":"Added blueprint mysql-blueprint","time":"2024-02-20T18:40:10.553588555Z"}
```
Testing issuing an invalid blueprint
```
(⎈|kind-local:default)➜  kanister git:(revert-k8s-upgrade-v0-29-x) k create -f ./examples/mysql/mysql-blueprint.yaml -n kanister
Error from server (Forbidden): error when creating "./examples/mysql/mysql-blueprint.yaml": admission webhook "blueprints.cr.kanister.io" denied the request: Invalid blueprint, Failed to validate action backup: Requested function {KubeTaskInvalid} has not been registered
```
Container logs for the above -
```
----------------------------------------------------------------------------  
Failed the 'validation of action backup' check.. ❌
{"level":"info","ts":"2024-02-20T18:41:29Z","logger":"admission","msg":"Validating blueprint","object":{"name":"mysql-blueprint","namespace":"kanister"},"namespace":"kanister","name":"mysql-blueprint","resource":{"group":"cr.kanister.io","version":"v1alpha1","resource":"blueprints"},"user":"kubernetes-admin","requestID":"109071e5-cafc-43d2-b8c1-408bb6fa1384"}
{"BlueprintName":"mysql-blueprint","File":"pkg/controller/controller.go","Function":"github.com/kanisterio/kanister/pkg/controller.(*Controller).onDeleteBlueprint","Line":300,"cluster_name":"94c9c8a4-f51d-46eb-a15c-019e6d6beabc","hostname":"kanister-kanister-operator-7df894bb95-gmlmr","level":"info","msg":"Deleted Blueprint ","time":"2024-02-20T18:44:02.815380997Z"}
{"level":"info","ts":"2024-02-20T18:44:13Z","logger":"admission","msg":"Validating blueprint","object":{"name":"mysql-blueprint","namespace":"kanister"},"namespace":"kanister","name":"mysql-blueprint","resource":{"group":"cr.kanister.io","version":"v1alpha1","resource":"blueprints"},"user":"kubernetes-admin","requestID":"c6413c18-a770-486b-89fa-10df3574dd13"}
```

### CODE REVIEW

**Review by viveksinghggits — CHANGES_REQUESTED** [review]

File: pkg/utils/utils.go [hunk 1]
```diff
@@ -114,3 +122,25 @@ func RoundUpDuration(t time.Duration) time.Duration {
 	}
 	return t.Round(time.Hour)
 }
+
+func getLogLevel() zapcore.Level {
```
[the remark this answers is not in the document]
Comment (mabhi) [reply pkg/utils/utils.go:126]: This is already being used in reposerver controller. https://github.com/kanisterio/kanister/blob/a76a5640f6f807b2d98b4887ce80eb16d7ab84cd/cmd/reposervercontroller/main.go#L156-L166
I have moved it to a place so as to refer it again

File: cmd/reposervercontroller/main.go [hunk 2]
```diff
@@ -138,11 +128,15 @@ func main() {
 	// we can use validating webhook for k8s server versions < 1.25
 	if k8sserverVersion.Major == "1" && minorVersion < minorK8sVersion {
 		if validatingwebhook.IsCACertMounted() {
-			hookServer := mgr.GetWebhookServer()
-			webhook := admission.WithCustomValidator(&crv1alpha1.RepositoryServer{}, &validatingwebhook.RepositoryServerValidator{})
+			hookServerOptions := webhook.Options{CertDir: validatingwebhook.WHCertsDir, Port: webhookServerPort}
+			hookServer := webhook.NewServer(hookServerOptions)
+			webhook := admission.WithCustomValidator(mgr.GetScheme(), &crv1alpha1.RepositoryServer{}, &validatingwebhook.RepositoryServerValidator{})
+			// registers a webhooks to a webhook server that gets ran by a controller manager.
```
[the remark this answers is not in the document]
Comment (mabhi) [reply cmd/reposervercontroller/main.go:134]: Done

File: pkg/resource/resource.go [hunk 3]
```diff
@@ -50,7 +50,7 @@ func CreateCustomResources(ctx context.Context, config *rest.Config) error {
 	return customresource.CreateCustomResources(*crCTX, resources)
 }
 
-func newOpKitContext(config *rest.Config) (*customresource.Context, error) {
+func newOpKitContext(config *rest.Config, ctx context.Context) (*customresource.Context, error) {
```
[the remark this answers is not in the document]
Comment (mabhi) [reply pkg/resource/resource.go:53]: Done

File: pkg/handler/handler.go [hunk 4]
```diff
@@ -62,17 +65,25 @@ func (*healthCheckHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 // RunWebhookServer starts the validating webhook resources for blueprint kanister resources
 func RunWebhookServer(c *rest.Config) error {
+	log.SetLogger(utils.GetLogger())
 	mgr, err := manager.New(c, manager.Options{})
 	if err != nil {
 		return errors.Wrapf(err, "Failed to create new webhook manager")
 	}
+	bpValidator := &validatingwebhook.BlueprintValidator{}
+	if err = bpValidator.InjectDecoder(admission.NewDecoder(mgr.GetScheme())); err != nil {
+		return errors.Wrapf(err, "Failed to inject decoder")
```
Comment (pavannd1) [inline pkg/handler/handler.go:75]:
```suggestion
		return errors.Wrap(err, "Failed to inject decoder")
```
Same as above

File: pkg/handler/handler.go [hunk 5]
```diff
@@ -62,17 +65,25 @@ func (*healthCheckHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 // RunWebhookServer starts the validating webhook resources for blueprint kanister resources
 func RunWebhookServer(c *rest.Config) error {
+	log.SetLogger(utils.GetLogger())
 	mgr, err := manager.New(c, manager.Options{})
 	if err != nil {
 		return errors.Wrapf(err, "Failed to create new webhook manager")
```
Comment (pavannd1) [inline pkg/handler/handler.go:71]:
```suggestion
		return errors.Wrap(err, "Failed to create new webhook manager")
```
No format specifiers used

File: pkg/handler/handler.go [hunk 6]
```diff
@@ -62,17 +65,25 @@ func (*healthCheckHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
-	hookServer.Register(whHandlePath, &webhook.Admission{Handler: &validatingwebhook.BlueprintValidator{}})
+	hookServerOptions := webhook.Options{CertDir: validatingwebhook.WHCertsDir}
+	hookServer := webhook.NewServer(hookServerOptions)
+	hookServer.Register(whHandlePath, &webhook.Admission{Handler: bpValidator})
 	hookServer.Register(healthCheckPath, &healthCheckHandler{})
 	hookServer.Register(metricsPath, promhttp.Handler())
 
-	hookServer.CertDir = validatingwebhook.WHCertsDir
+	if err := mgr.Add(hookServer); err != nil {
+		return errors.Wrapf(err, "Failed to add new webhook server")
```
Comment (pavannd1) [inline pkg/handler/handler.go:85]:
```suggestion
		return errors.Wrap(err, "Failed to add new webhook server")
```
Same as above

File: pkg/validatingwebhook/blueprint_handler.go [hunk 7]
```diff
@@ -3,13 +3,13 @@ package validatingwebhook
...
 	"context"
 	"fmt"
-	"net/http"
 
 	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 
 	kanister "github.com/kanisterio/kanister/pkg"
 	crv1alpha1 "github.com/kanisterio/kanister/pkg/apis/cr/v1alpha1"
 	"github.com/kanisterio/kanister/pkg/blueprint/validate"
+	"net/http"
```
Comment (pavannd1) [inline pkg/validatingwebhook/blueprint_handler.go:12]: This change is not required. `net/http` should be moved back to the list above

Comment (mabhi) [reply pkg/validatingwebhook/blueprint_handler.go:12]: Done

File: pkg/handler/handler.go (hunk 4 above)
Comment (mabhi) [reply pkg/handler/handler.go:75]: Done

File: pkg/handler/handler.go (hunk 5 above)
Comment (mabhi) [reply pkg/handler/handler.go:71]: Done

File: pkg/handler/handler.go (hunk 6 above)
Comment (mabhi) [reply pkg/handler/handler.go:85]: Done

File: cmd/reposervercontroller/main.go [hunk 8]
```diff
@@ -20,31 +20,29 @@ import (
...
 	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 
 	crv1alpha1 "github.com/kanisterio/kanister/pkg/apis/cr/v1alpha1"
 	"github.com/kanisterio/kanister/pkg/controllers/repositoryserver"
-	"github.com/kanisterio/kanister/pkg/log"
 	"github.com/kanisterio/kanister/pkg/resource"
 	"github.com/kanisterio/kanister/pkg/validatingwebhook"
+
 	//+kubebuilder:scaffold:imports
+	"github.com/kanisterio/kanister/pkg/utils"
```
Comment (PrasadG193) [inline cmd/reposervercontroller/main.go:40]: Please add it to the correct import groups. Please avoid adding unnecessary new lines changes

File: pkg/utils/utils.go [hunk 9]
```diff
@@ -114,3 +122,25 @@ func RoundUpDuration(t time.Duration) time.Duration {
...
+		return defaultLogLevel
+	}
+	level, err := zapcore.ParseLevel(logLevel)
+	if err != nil {
+		return defaultLogLevel
+	}
+	return level
+}
+
+func GetLogger() logr.Logger {
```
Comment (PrasadG193) [inline pkg/utils/utils.go:138]: We shouldn't move this to the utility function. This is not the default logger we use in Kanister.
It would mislead developers.

Comment (PrasadG193) [reply pkg/utils/utils.go:138]: Please check if we can use default logger from `github.com/kanisterio/kanister/pkg/log` instead of creating this new type of logger.

File: cmd/reposervercontroller/main.go [hunk 10]
```diff
@@ -64,22 +62,15 @@ func main() {
...
-		Level: logLevel,
-	}
 
 	// Disable metrics server
 	metricsAddr = "0"
-	opts.BindFlags(flag.CommandLine)
 	flag.Parse()
-	logger := zap.New(zap.UseFlagOptions(&opts))
-	ctrl.SetLogger(logger)
+	ctrl.SetLogger(utils.GetLogger())
```
Comment (PrasadG193) [inline cmd/reposervercontroller/main.go:69]: We should also avoid making unrelated changes to K8s upgrade in the same PR. Unrelated improvements can be done as a follow-up PR.
Added more comments on the `GetLogger` method below

File: docker/postgres-kanister-tools/Dockerfile [hunk 11]
```diff
@@ -4,15 +4,15 @@ ARG TOOLS_IMAGE
...
 
 ENV DEBIAN_FRONTEND noninteractive
 
 USER root
 
 RUN apt-get update && apt-get -y install curl python3 groff less jq python3-pip && \
-    pip3 install --break-system-packages --upgrade pip && \
-    pip3 install --break-system-packages --upgrade awscli && \
+    pip3 install --upgrade pip && \
+    pip3 install --upgrade awscli && \
```
Comment (PrasadG193) [inline docker/postgres-kanister-tools/Dockerfile:15]: Is it related to K8s dep upgrade? If not, please avoid making changes in the same PR.

File: examples/postgres-RDS/Dockerfile [hunk 12]
```diff
@@ -5,8 +5,8 @@ ENV DEBIAN_FRONTEND noninteractive
 USER root
 
 RUN apt-get update && apt-get -y install curl python3-pip && \
-    pip3 install --break-system-packages --upgrade pip && \
-    pip3 install --break-system-packages --upgrade awscli && \
+    pip3 install --upgrade pip && \
+    pip3 install --upgrade awscli && \
```
Comment (PrasadG193) [inline examples/postgres-RDS/Dockerfile:9]: Same as above

File: pkg/blockstorage/vmware/vmware.go [hunk 13]
```diff
@@ -257,7 +257,8 @@ var reVslmSyncFaultFatal = regexp.MustCompile("Change tracking invalid or disk i
 func (p *FcdProvider) SnapshotCreate(ctx context.Context, volume blockstorage.Volume, tags map[string]string) (*blockstorage.Snapshot, error) {
 	var res types.AnyType
 	description := generateSnapshotDescription(tags)
-	err := wait.PollImmediate(time.Second, defaultRetryLimit, func() (bool, error) {
+
```
Comment (PrasadG193) [inline pkg/blockstorage/vmware/vmware.go:260]:
```suggestion
```
Unnecessary whitespace change

File: docker/postgres-kanister-tools/Dockerfile (hunk 11 above)
Comment (mabhi) [reply docker/postgres-kanister-tools/Dockerfile:15]: Understood, this is not coming from my changes but because of merging master changes.

File: examples/postgres-RDS/Dockerfile (hunk 12 above)
Comment (mabhi) [reply examples/postgres-RDS/Dockerfile:9]: Same as above.

File: docker/postgres-kanister-tools/Dockerfile (hunk 11 above)
Comment (PrasadG193) [reply docker/postgres-kanister-tools/Dockerfile:15]: Please rebase with master, the PR should not contain other commits. Lets fix that

File: cmd/reposervercontroller/main.go (hunk 8 above)
Comment (mabhi) [reply cmd/reposervercontroller/main.go:40]: Done

File: pkg/utils/utils.go (hunk 9 above)
Comment (mabhi) [reply pkg/utils/utils.go:138]: Default logger from [github.com/kanisterio/kanister/pkg/log](http://github.com/kanisterio/kanister/pkg/log) ,   does not comply with the log.logger. Anything else I can use here

File: cmd/reposervercontroller/main.go (hunk 10 above)
Comment (mabhi) [reply cmd/reposervercontroller/main.go:69]: Reverted/

File: pkg/blockstorage/vmware/vmware.go (hunk 13 above)
Comment (mabhi) [reply pkg/blockstorage/vmware/vmware.go:260]: Done

File: pkg/utils/utils.go (hunk 9 above)
Comment (mabhi) [reply pkg/utils/utils.go:138]: Using `log.SetLogger(logr.New(log.NullLogSink{}))` from `sigs.k8s.io/controller-runtime/pkg/log` package. 

In the default implementation logger is being initialised in controller-runtime by using `NullLogSink` in case of nothing is set explicitly. As we want to keep the behaviour similar to what it used to be earlier and also to get rid of the warning, we use similar approach.

File: docker/postgres-kanister-tools/Dockerfile (hunk 11 above)
Comment (mabhi) [reply docker/postgres-kanister-tools/Dockerfile]: Done

File: pkg/utils/utils.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mabhi) [reply pkg/utils/utils.go]: This assignment is necessary to avoid memory leaks. More on this https://github.com/kubernetes-sigs/controller-runtime/issues/1655

[the remark this answers is not in the document]
Comment (mabhi) [reply pkg/utils/utils.go]: https://github.com/kanisterio/kanister/pull/2681#discussion_r1500532805 This change was reverted. Not moving this to any common place any more.

**Review by PrasadG193 — APPROVED** [review]
LGTM

**Review by viveksinghggits — APPROVED** [review]

File: pkg/blockstorage/vmware/vmware.go [hunk 14]
```diff
@@ -335,7 +335,7 @@ func (p *FcdProvider) SnapshotDelete(ctx context.Context, snapshot *blockstorage
 	if err != nil {
 		return errors.Wrap(err, "Cannot infer volume ID from full snapshot ID")
 	}
-	return wait.PollImmediate(time.Second, defaultRetryLimit, func() (bool, error) {
+	return wait.PollUntilContextTimeout(ctx, time.Second, defaultRetryLimit, false, func(context.Context) (bool, error) {
```
Comment (pavannd1) [inline pkg/blockstorage/vmware/vmware.go:338]: Does the internal `func` need a named argument for context? This is done above on line 260 - it's named `innerCtx`
